src/transaction_builder.js

```diff
@@ -437,65 +405,101 @@ TransactionBuilder.prototype.sign = function (index, keyPair, redeemScript, hash
+        var prevOutScriptScriptHash = bscript.decompile(input.prevOutScript)[1]
+        if (!bufferEquals(prevOutScriptScriptHash, redeemScriptHash)) throw new Error('Inconsistent hash160(RedeemScript)')
 
-      // if we don't have a prevOutScript, generate a P2SH script
-      if (!input.prevOutScript) {
-        input.prevOutScript = bscript.scriptHashOutput(bcrypto.hash160(redeemScript))
+      // or, we don't have a prevOutScript, so generate a P2SH script
+      } else {
+        input.prevOutScript = bscript.scriptHashOutput(redeemScriptHash)
         input.prevOutType = 'scripthash'
       }
 
-      input.pubKeys = extracted.pubKeys
+      var expanded = expandOutput(redeemScript, kpPubKey)
+      if (!expanded) throw new Error('RedeemScript not supported "' + bscript.toASM(redeemScript) + '"')
+
+      input.pubKeys = expanded.pubKeys
       input.redeemScript = redeemScript
-      input.redeemScriptType = extracted.scriptType
-      input.signatures = extracted.pubKeys.map(function () { return undefined })
+      input.redeemScriptType = expanded.scriptType
+      input.signatures = expanded.signatures
+
+    // no redeemScript
     } else {
       // pay-to-scriptHash is not possible without a redeemScript
       if (input.prevOutType === 'scripthash') throw new Error('PrevOutScript is P2SH, missing redeemScript')
 
-      // if we don't have a scriptType, assume pubKeyHash otherwise
-      if (!input.scriptType) {
+      // if we don't have a scriptType, assume pubKeyHash
+      if (!input.prevOutType) {
         input.prevOutScript = bscript.pubKeyHashOutput(bcrypto.hash160(kpPubKey))
         input.prevOutType = 'pubkeyhash'
 
         input.pubKeys = [kpPubKey]
-        input.scriptType = input.prevOutType
         input.signatures = [undefined]
       } else {
         // throw if we can't sign with it
-        if (!input.pubKeys || !input.signatures) throw new Error(input.scriptType + ' not supported')
+        if (!input.pubKeys || !input.signatures) throw new Error(input.prevOutType + ' not supported')
       }
     }
 
     input.hashType = hashType
   }
 
   // ready to sign?
-  var signatureScript = input.redeemScript || input.prevOutScript
-  var signatureHash = this.tx.hashForSignature(index, signatureScript, hashType)
+  var hashScript = input.redeemScript || input.prevOutScript
+  var signatureHash = this.tx.hashForSignature(index, hashScript, hashType)
 
   // enforce in order signing of public keys
   var valid = input.pubKeys.some(function (pubKey, i) {
     if (!bufferEquals(kpPubKey, pubKey)) return false
     if (input.signatures[i]) throw new Error('Signature already exists')
 
     input.signatures[i] = keyPair.sign(signatureHash)
-
     return true
   })
 
   if (!valid) throw new Error('Key pair cannot sign for this input')
 }
 
+TransactionBuilder.prototype.__fixMultisigOrder = function (transaction, vin) {
```

Comment:
This could be `function fixMultisigOrder (vin) {` if I did a iteration in `fromTransaction`... but maybe another time